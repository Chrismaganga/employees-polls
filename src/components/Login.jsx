// import React, { useState } from 'react';
// import { useSelector, useDispatch } from 'react-redux';
// import { setAuthedUser } from '../slices/authSlice';
// import { useNavigate } from 'react-router-dom';

// function Login() {
//   const users = useSelector(state => state.users.users);
//   const usersStatus = useSelector(state => state.users.status);
//   const [selectedUser, setSelectedUser] = useState('');
//   const dispatch = useDispatch();
//   const navigate = useNavigate();

//   const handleSubmit = (e) => {
//     e.preventDefault();
//     if (selectedUser) {
//       dispatch(setAuthedUser(selectedUser));
//       navigate('/');
//     }
//   };

//   if (usersStatus === 'loading') {
//     return <div>Loading...</div>;
//   }

//   return (
//     <div className="login-container">
//       <h2>Welcome to Employee Polls</h2>
//       <form onSubmit={handleSubmit}>
//         <label htmlFor="user">Sign In</label>
//         <select
//           id="user"
//           value={selectedUser}
//           onChange={(e) => setSelectedUser(e.target.value)}
//         >
//           <option value="" disabled>
//             Select User
//           </option>
//           {Object.values(users).map(user => (
//             <option key={user.id} value={user.id}>
//               {user.name}
//             </option>
//           ))}
//         </select>
//         <button type="submit" disabled={!selectedUser}>
//           Login
//         </button>
//       </form>
//     </div>
//   );
// }

// export default Login;
import React, { useState } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { setAuthedUser } from '../slices/authSlice';
import { useNavigate } from 'react-router-dom';

function Login() {
  const users = useSelector(state => state.users.users);
  const usersStatus = useSelector(state => state.users.status);
  const [selectedUser, setSelectedUser] = useState('');
  const dispatch = useDispatch();
  const navigate = useNavigate();

  const handleSubmit = (e) => {
    e.preventDefault();
    if (selectedUser) {
      dispatch(setAuthedUser(selectedUser));
      navigate('/');
    }
  };

  if (usersStatus === 'loading') {
    return <div>Loading...</div>;
  }

  return (
    <div className="login-container">
      <h2>Welcome to Employee Polls</h2>
      <form onSubmit={handleSubmit}>
        <label htmlFor="user">Sign In</label>
        <select
          id="user"
          value={selectedUser}
          onChange={(e) => setSelectedUser(e.target.value)}
        >
          <option value="" disabled>
            Select User
          </option>
          {Object.values(users).map(user => (
            <option key={user.id} value={user.id}>
              {user.name}
            </option>
          ))}
        </select>

        {selectedUser && (
          <div className="avatar-preview">
            <img
              src={users[selectedUser].avatarURL || "/home/chris/voting/src/images/placeholder.png"}
              alt={`${users[selectedUser].name}'s avatar`}
              style={{ width: '100px', height: '100px' }}
            />
          </div>
        )}
        
        <button type="submit" disabled={!selectedUser}>
          Login
        </button>
      </form>
    </div>
  );
}

export default Login;
