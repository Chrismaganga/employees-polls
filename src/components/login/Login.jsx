import React, { useState } from 'react';
import { useNavigate } from 'react-router-dom';
import { useDispatch, useSelector } from 'react-redux';
import { setAuthedUser, setUser } from '../../slices/authSlice';
import './Login.css';

function Login() {
  const [username, setUsername] = useState('');
  const [password, setPassword] = useState('');
  const [error, setError] = useState('');
  const navigate = useNavigate();
  const dispatch = useDispatch();
  const users = useSelector(state => state.users.users);

  const handleSubmit = (e) => {
    e.preventDefault();
    setError('');

    // Find user in the users list
    const user = Object.values(users).find(
      u => u.id === username && u.password === password
    );

    if (user) {
      dispatch(setAuthedUser(username));
      dispatch(setUser(user));
      navigate('/');
    } else {
      setError('Invalid username or password');
    }
  };

  return (
    <div className="login-container">
      <div className="login-box">
        <h2>Employee Polls</h2>
        <form onSubmit={handleSubmit}>
          {error && <div className="error-message">{error}</div>}
          
          <div className="input-group">
            <label htmlFor="username">Username</label>
            <input
              type="text"
              id="username"
              value={username}
              onChange={(e) => setUsername(e.target.value)}
              placeholder="Enter your username"
              required
            />
          </div>

          <div className="input-group">
            <label htmlFor="password">Password</label>
            <input
              type="password"
              id="password"
              value={password}
              onChange={(e) => setPassword(e.target.value)}
              placeholder="Enter your password"
              required
            />
          </div>

          <button type="submit" className="login-button">
            Login
          </button>
        </form>
      </div>
    </div>
  );
}

export default Login;
