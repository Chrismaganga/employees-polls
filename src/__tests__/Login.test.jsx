import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import configureStore from 'redux-mock-store';
import Login from '../components/login/Login';

const mockStore = configureStore([]);

describe('Login Component', () => {
  let store;

  beforeEach(() => {
    store = mockStore({
      users: {
        users: {
          testuser: {
            id: 'testuser',
            password: 'testpass',
            name: 'Test User',
            avatarURL: 'test-avatar.jpg'
          }
        },
        status: 'succeeded'
      }
    });
    store.dispatch = jest.fn();
  });

  it('renders correctly', () => {
    const { container } = render(
      <Provider store={store}>
        <BrowserRouter>
          <Login />
        </BrowserRouter>
      </Provider>
    );
    expect(container).toMatchSnapshot();
  });

  it('handles form submission correctly', () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <Login />
        </BrowserRouter>
      </Provider>
    );

    fireEvent.change(screen.getByLabelText('Username'), {
      target: { value: 'testuser' }
    });
    fireEvent.change(screen.getByLabelText('Password'), {
      target: { value: 'testpass' }
    });
    fireEvent.click(screen.getByText('Login'));

    expect(store.dispatch).toHaveBeenCalled();
  });

  it('shows error for invalid credentials', () => {
    const alertMock = jest.spyOn(window, 'alert').mockImplementation(() => {});
    
    render(
      <Provider store={store}>
        <BrowserRouter>
          <Login />
        </BrowserRouter>
      </Provider>
    );

    fireEvent.change(screen.getByLabelText('Username'), {
      target: { value: 'wronguser' }
    });
    fireEvent.change(screen.getByLabelText('Password'), {
      target: { value: 'wrongpass' }
    });
    fireEvent.click(screen.getByText('Login'));

    expect(alertMock).toHaveBeenCalledWith('Invalid username or password');
    alertMock.mockRestore();
  });
});
