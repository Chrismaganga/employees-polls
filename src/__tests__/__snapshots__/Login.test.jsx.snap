// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Component renders correctly 1`] = `
<div>
  <div
    class="login-container"
  >
    <div
      class="login-box"
    >
      <h2>
        Employee Polls
      </h2>
      <form>
        <div
          class="input-group"
        >
          <label
            for="username"
          >
            Username
          </label>
          <input
            id="username"
            placeholder="Enter your username"
            required=""
            type="text"
            value=""
          />
        </div>
        <div
          class="input-group"
        >
          <label
            for="password"
          >
            Password
          </label>
          <input
            id="password"
            placeholder="Enter your password"
            required=""
            type="password"
            value=""
          />
        </div>
        <button
          class="login-button"
          type="submit"
        >
          Login
        </button>
      </form>
    </div>
  </div>
</div>
`;
